manifests/certonly.pp

```diff
@@ -56,62 +67,95 @@
   Variant[Integer[0,23], String, Array]     $cron_hour            = fqdn_rand(24, $title),
   Variant[Integer[0,59], String, Array]     $cron_minute          = fqdn_rand(60, fqdn_rand_string(10, $title)),
   Stdlib::Unixpath                          $config_dir           = $letsencrypt::config_dir,
+  Array[String[1]]                          $pre_hook_commands    = [],
+  Array[String[1]]                          $post_hook_commands   = [],
+  Array[String[1]]                          $deploy_hook_commands = [],
 ) {
 
   if $plugin == 'webroot' and empty($webroot_paths) {
     fail("The 'webroot_paths' parameter must be specified when using the 'webroot' plugin")
   }
 
+  # Wildcard-less title for use in file paths
+  $title_nowc = regsubst($title, '^\*\.', '')
+
   if ($custom_plugin) {
-    $command_start = "${letsencrypt_command} --text --agree-tos --non-interactive certonly --rsa-key-size ${key_size} "
+    $default_args = "--text --agree-tos --non-interactive certonly --rsa-key-size ${key_size}"
   } else {
-    $command_start = "${letsencrypt_command} --text --agree-tos --non-interactive certonly --rsa-key-size ${key_size} -a ${plugin} "
+    $default_args = "--text --agree-tos --non-interactive certonly --rsa-key-size ${key_size} -a ${plugin}"
   }
 
   case $plugin {
 
     'webroot': {
-      $_command_domains = zip($domains, $webroot_paths).map |$domain| {
+      $_plugin_args = zip($domains, $webroot_paths).map |$domain| {
         if $domain[1] {
           "--webroot-path ${domain[1]} -d ${domain[0]}"
         } else {
           "-d ${domain[0]}"
         }
       }
-      $command_domains = join([ "--cert-name ${title}", ] + $_command_domains, ' ')
+      $plugin_args = ["--cert-name ${title}"] + $_plugin_args
     }
 
     'dns-rfc2136': {
       require letsencrypt::plugin::dns_rfc2136
-      $dns_args = [
+      $plugin_args = [
         "--cert-name ${title} -d",
         join($domains, ' -d '),
         "--dns-rfc2136-credentials ${letsencrypt::plugin::dns_rfc2136::config_dir}/dns-rfc2136.ini",
         "--dns-rfc2136-propagation-seconds ${letsencrypt::plugin::dns_rfc2136::propagation_seconds}",
       ]
-      $command_domains = join($dns_args, ' ')
     }
 
     default: {
-      $_command_domains = join($domains, ' -d ')
-      $command_domains  = "--cert-name ${title} -d ${_command_domains}"
+      $_plugin_args = join($domains, ' -d ')
+      $plugin_args  = "--cert-name ${title} -d ${_plugin_args}"
     }
 
   }
 
-  if empty($additional_args) {
-    $command_end = undef
-  } else {
-    # ['',] adds an additional whitespace in the front
-    $command_end = join(['',] + $additional_args, ' ')
+  $hook_args = ['pre', 'post', 'deploy'].map | String $type | {
+    $commands = getvar("${type}_hook_commands")
+    if (!empty($commands)) {
+      $hook_file = "${letsencrypt::renew::config_dir}/renewal-hooks-puppet/${type}-${title_nowc}.sh"
```

Comment:
@bastelfreak @Dan33l This config_dir reference has me a little stumped in rspec. I wanted to be sure it uses the same directory as `certbot::renew` in case the user chooses a custom path just for that class, but how do I supply it to rspec when the `letsencrypt` class contains it? If I add a pre_condition for `letsencrypt::renew`, I'll get a duplicate class declaration.

Reply:
I have the samep roblem in `letsencrypt::renew` with references to `letsencrypt` variables. If I try to supply them with a class declaration for `letsencrypt` in a `pre_condition`, I'll get a duplicate class error.

Reply:
Solved this by defaulting the vars in params, declaring them in `letsencrypt`, and referencing them from `letsencrypt::renew` to make the needed vars available without extra declarations. I'm spoiled by Hiera.